### PR TITLE
added user class and ensure new message from sockets is sent to the r…

### PR DESCRIPTION
**Pull request by connieyu18**
Created another branch- for making sure that only the user who wants a message will receive it , rather than broadcasting to the entire network
- added User Class which stored the user and socket id as a object instead of array

### CODE REVIEW

File: server/onlineUsers.js [hunk 1]
```diff
@@ -1,2 +1,42 @@
...
+    }
+    return null;
+  }
+
+  delete(userId) {
+    delete this.users[userId];
+  }
+}
+
+const onlineUsers = new OnlineUsers();
```
Comment (kuperavv) [inline server/onlineUsers.js:43]: Great, this is a nice structure to handle the online users.

File: server/onlineUsers.js [hunk 2]
```diff
@@ -1,2 +1,42 @@
...
+  }
+
+  // checks if userId exists
+  includes(userId) {
+    return userId in this.users;
+  }
+
+  // returns true if added, false if not added
+  addUser(userId, socketId) {
+    if (this.includes(userId)) return false;
```
Comment (kuperavv) [inline server/onlineUsers.js]: This may prevent the same user to be logged into multiple devices simultaneously.

File: server/onlineUsers.js [hunk 3]
```diff
@@ -1,2 +1,42 @@
...
+    this.users[userId] = {
+      socketId: socketId,
+    };
+    return true;
+  }
+
+  // returns the user structure
+  // null if it does not exist
+  getUser(userId) {
+    if (this.includes(userId)) {
```
Comment (kuperavv) [inline server/onlineUsers.js]: Should this be  (userId in this.users) or similar?  I haven't this this.includes on an object before (typically an array method).

File: server/bin/www [hunk 4]
```diff
@@ -34,23 +34,29 @@ const io = require("socket.io")(server);
...
   socket.on("new-message", (data) => {
-    socket.broadcast.emit("new-message", {
+    let user = onlineUsers.getUser(data.recipientId);
+    if (!user || !user.socketId) {
+      console.error("User or user socket does not exist");
+      return;
+    }
+
+    // send to specific user's socket
+    io.to(user.socketId).emit("new-message", {
```
Comment (kuperavv) [inline server/bin/www:51]: Nice fix!

**Review by kuperavv — COMMENTED** [review]
Nice work!  I had a question about this.includes syntax (either I haven't seen it before, or it's a bug).

File: server/bin/www [hunk 5]
```diff
@@ -34,23 +34,29 @@ const io = require("socket.io")(server);
...
-      onlineUsers.push(id);
+      onlineUsers.addUser(id, socket.id);
     }
     // send the user who just went online to everyone else who is already online
     socket.broadcast.emit("add-online-user", id);
   });
 
   socket.on("new-message", (data) => {
-    socket.broadcast.emit("new-message", {
+    let user = onlineUsers.getUser(data.recipientId);
```
Comment (ThinkTink) [inline server/bin/www:44]: We can use const here - as a general rule of thumb, it's always best to use const over let when you can

File: server/onlineUsers.js [hunk 6]
```diff
@@ -1,2 +1,44 @@
...
+      socketId: socketId,
+    };
+    return true;
+  }
+
+  // returns the user structure
+  // null if it does not exist
+  getUser(userId) {
+    if (this.user.includes(userId)) {
+      if (this.users.includes(userId)) {
```
Comment (ThinkTink) [inline server/onlineUsers.js:31]: is this meant to be like this?

**Review by ThinkTink — APPROVED** [review]
:+1:

**Pull request merged**